src/storage/txn/scheduler.rs

```diff
@@ -11,92 +11,497 @@
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Prewrite { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Commit { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Cleanup { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+            Command::Rollback { ref mut callback, .. } => {
+                callback.take().unwrap()(Err(err.into()));
+            }
+        }
+    }
+
+    fn finish_with_err(&mut self, cid: u64, err: Error) {
+        self.process_failed_cmd(cid, err);
+        self.finish_cmd(cid);
+    }
+
+    fn extract_context(&self, cid: u64) -> &Context {
+        let ctx: &RunningCtx = self.cmd_ctxs.get(&cid).unwrap();
+        match ctx.cmd {
+            Command::Get { ref ctx, .. } |
+            Command::BatchGet { ref ctx, .. } |
+            Command::Scan { ref ctx, .. } |
+            Command::Prewrite { ref ctx, .. } |
+            Command::Commit { ref ctx, .. } |
+            Command::CommitThenGet { ref ctx, .. } |
+            Command::Cleanup { ref ctx, .. } |
+            Command::Rollback { ref ctx, .. } |
+            Command::RollbackThenGet { ref ctx, .. } => {
+                ctx
+            }
         }
-        Command::Prewrite { ctx, mutations, primary, start_ts, callback } => {
-            callback(match store.prewrite(ctx, mutations, primary, start_ts) {
-                Ok(mut results) => {
-                    Ok(results.drain(..).map(|x| x.map_err(::storage::Error::from)).collect())
+    }
+
+    fn register_report_tick(&self, event_loop: &mut EventLoop<Self>) {
+        if let Err(e) = register_timer(event_loop, Tick::ReportStatistic, REPORT_STATISTIC_INTERVAL) {
+            error!("register report statistic err: {:?}", e);
+        };
+    }
+
+    fn on_report_staticstic_tick(&self, event_loop: &mut EventLoop<Self>) {
+        info!("all running cmd count = {}", self.cmd_ctxs.len());
+
+        self.register_report_tick(event_loop);
+    }
+
+    fn received_new_cmd(&mut self, cmd: Command) {
+        let cid = self.gen_id();
+
+        if cmd.readonly() {
+            // readonly commands don't need latch
+            let ctx = RunningCtx::new(cid, cmd, vec![]);
+            self.save_cmd_context(cid, ctx);
+            self.get_snapshot(cid);
+        } else {
+            // write command needs acquire latches first, or this command will not
+            // be handled currently, it will be woken up by other command
+            let slots = self.calc_latches_slots(&cmd);
+            let ctx = RunningCtx::new(cid, cmd, slots);
+            self.save_cmd_context(cid, ctx);
+
+            let all_acquired = self.acquire_latches(cid);
+            if all_acquired {
+                self.get_snapshot(cid);
+            }
```

Comment:
Above 4 lines are the same as `wakeup_cmd`. Maybe rename the function then call it here?